packages/docusaurus/src/commands/swizzle/actions.ts

```diff
@@ -0,0 +1,156 @@
+  if (options.wrap) {
+    return 'wrap';
+  }
+  if (options.eject) {
+    return 'eject';
+  }
+  return askSwizzleAction(componentConfig);
+}
+
+export type ActionParams = {
+  siteDir: string;
+  themePath: string;
+  componentName: string;
+};
+
+export type ActionResult = {
+  createdFiles: string[];
+};
+
+async function isDir(dirPath: string): Promise<boolean> {
+  return (
+    (await fs.pathExists(dirPath)) && (await fs.stat(dirPath)).isDirectory()
+  );
+}
+
+export async function eject({
+  siteDir,
+  themePath,
+  componentName,
+}: ActionParams): Promise<ActionResult> {
+  const fromPath = path.join(themePath, componentName);
+
+  const isDirectory = await isDir(fromPath);
+
+  const globIgnore = ['**/*.{story,stories,test,tests}.{js,jsx,ts,tsx}'];
+
+  const globPattern = isDirectory
+    ? // do we really want to copy all components?
+      path.join(fromPath, '*')
+    : `${fromPath}.*`;
+
+  const globPatternPosix = posixPath(globPattern);
+
+  const filesToCopy = await Globby(globPatternPosix, {
+    ignore: globIgnore,
+  });
+
+  if (filesToCopy.length === 0) {
+    // This should never happen
+    throw new Error(
+      logger.interpolate`No files to copy from path=${fromPath} with glob code=${globPatternPosix}`,
+    );
+  }
+
+  const toPath = isDirectory
+    ? path.join(siteDir, THEME_PATH, componentName)
+    : path.join(siteDir, THEME_PATH);
+
+  await fs.ensureDir(toPath);
+
+  async function copyFile(sourceFile: string) {
+    const fileName = path.basename(sourceFile);
+    const targetFile = path.join(toPath, fileName);
+    try {
+      await fs.copy(sourceFile, targetFile, {overwrite: true});
+    } catch (err) {
+      throw new Error(
+        logger.interpolate`Could not copy file from ${sourceFile} to ${targetFile}`,
+      );
+    }
+    return targetFile;
+  }
+
+  const createdFiles = await Promise.all(filesToCopy.map(copyFile));
+
+  return {createdFiles};
+}
+
+export async function wrap({
+  siteDir,
+  themePath,
+  componentName: themeComponentName,
+  typescript,
+  importType = 'original',
+}: ActionParams & {
+  typescript: boolean;
+  importType?: 'original' | 'init';
+}): Promise<ActionResult> {
+  const isDirectory = await isDir(path.join(themePath, themeComponentName));
+
+  // Top/Parent/ComponentName => ComponentName
+  const componentName = _.last(themeComponentName.split('/'));
+  const wrapperComponentName = `${componentName}Wrapper`;
+
+  const wrapperFileName = `${themeComponentName}${isDirectory ? '/index' : ''}${
+    typescript ? '.tsx' : '.js'
+  }`;
+
+  await fs.ensureDir(path.resolve(siteDir, THEME_PATH));
+
+  const toPath = path.resolve(siteDir, THEME_PATH, wrapperFileName);
+
+  const content = typescript
+    ? `import React, {ComponentProps} from 'react';
+import type ${componentName}Type from '@theme/${themeComponentName}';
+import ${componentName} from '@theme-${importType}/${themeComponentName}';
+
+type Props = ComponentProps<typeof ${componentName}Type>
```

Comment:
Any particular reason for using this structure instead of `import {Props}`? I think there's zero place where we don't export the props as `interface Props`.

Reply:
We can't guarantee all TS theme authors will always export a type named Props, and not 100% sure we should ask them to do so

It is probably not the case in our ecosystem already today

Reply:
Ah, I see. It's hard to design, but this is almost an anti-pattern. Probably good enough for now